More Dict and Set examples

aminya opened the pull request:
Ready to get merged...

aminya committed:
Adding documentation and examples for Set

aminya committed:
More examples for Dict creation

get doc for d[x] syntax

aminya committed:
sets doc rewording

aminya committed:
Move set examples to docstring

aminya committed:
Adding speed warning

aminya committed:
Dict getting and changing value doc

Fixed julia guessing type

space

Fixes https://github.com/JuliaLang/julia/pull/33936#discussion_r350047741

remove space

https://github.com/JuliaLang/julia/pull/33936#discussion_r351792034

Fixed bug

Fixes https://github.com/JuliaLang/julia/pull/33936#discussion_r350048362

jonas-schulze commented on base/dict.jl:
````diff
@@ -69,11 +69,61 @@ Dict{String,Int64} with 2 entries:
   "A" => 1
 ```
+
+After creation of a Dict you can use `d[x]`for getting the values and `d[x]=y` for changing the values.
````
There is a space missing: `d[x]`for

jonas-schulze commented on base/dict.jl:
````diff
@@ -69,11 +69,61 @@ Dict{String,Int64} with 2 entries:
+First, you need to create an empty dictionary using:
+```jldoctest
+julia> Dict{Int64,Any}()  # type specified
+Dict{String,Any} with 0 entries
````
`Int64` becomes `String` 🤔

aminya replied:
Good catch! I wanted to edit something, but I forgot to edit the answer.

jonas-schulze commented on doc/src/base/collections.md:
````diff
@@ -234,6 +234,51 @@ Partially implemented by:
+
+To see if something is in the set, you can use [`in`](@Ref):
+```julia
+julia> in("green", colors)
````
Maybe the infix notation looks more familiar: `"green" in colors`

jonas-schulze commented:
The way you print the sets might be incompatible with #33300.

aminya committed:
infix notation for in

Fixes https://github.com/JuliaLang/julia/pull/33936#discussion_r350051581

aminya commented:
> 
> 
> The way you print the sets might be incompatible with #33300.

I should check with nightly then.

jonas-schulze commented on base/dict.jl:
````diff
@@ -69,11 +69,61 @@ Dict{String,Int64} with 2 entries:
 Alternatively, a sequence of pair arguments may be passed.
 
 ```jldoctest
````
To fix part of the [failing doctests](https://build.julialang.org/#/builders/148/builds/3366/steps/3/logs/stdio), you could add [a label to the code blocks](https://juliadocs.github.io/Documenter.jl/stable/man/doctests/#Preserving-Definitions-Between-Blocks-1) that should share their context:

```suggestion
```jldoctest dictexamples
```

jonas-schulze commented on base/dict.jl:
````diff
@@ -69,11 +69,61 @@ Dict{String,Int64} with 2 entries:
 ```
+
+After creation of a Dict you can use `d[x]` for getting the values and `d[x]=y` for changing the values.
+```jldoctest
````
```suggestion
```jldoctest dictexamples
```

jonas-schulze commented on base/dict.jl:
````diff
@@ -69,11 +69,61 @@ Dict{String,Int64} with 2 entries:
+  2 => 4
+  3 => 9
+  1 => 1
+ ```
````
This causes another doc test to fail:

````suggestion
```
````

jonas-schulze replied:
Maybe opening and closing code fences must have the same indentation. 🤔

jonas-schulze reviewed (commented):
I like the additional examples. 🙂

I just noticed that your examples for sets are no `jldoctest`s. If that is intentional (as to not rely on a particular order of the elements when printing; there might even be another issue/PR for that) I think that's fine.

jonas-schulze commented on base/dict.jl:
````diff
@@ -69,11 +69,61 @@ Dict{String,Int64} with 2 entries:
 Dict{String,Int64} with 2 entries:
   "B" => 2
   "A" => 1
 ```
````
````suggestion
```
````

aminya committed:
jldoctest context

aminya committed:
Filling empty sets using push!

Adding jldoctest setexamples

Set crossreference

aminya committed:
Looping through dict example

aminya commented:
This is ready to get merged.

KristofferC commented on base/set.jl:
```diff
@@ -17,6 +17,37 @@ Set() = Set{Any}()
+julia> intContainer = Set{Int64}() # type specified
+Set(Int64[])
+
+julia> push!(intContainer, 2)
```
Needs to adhere to style guide for variables (same for other places).

KristofferC commented on base/dict.jl:
```diff
@@ -68,12 +68,65 @@ Dict{String,Int64} with 2 entries:
+    println("Animal group-$key are $(value[1]) and we have $(value[2]) number of them")
+end
+
+Animal group-2 are fish and we have 11185965267091429831 number of them
```
> Animal group-2 are fish and we have 11185965267091429831 number of them

That seems like a lot of fish...

aminya replied:
That is
eleven quintillion, 
one hundred eighty-five quadrillion, 
nine hundred sixty-five trillion, 
two hundred sixty-seven billion, 
ninety-one million, 
four hundred twenty-nine thousand, 
eight hundred thirty-one 
🐟

aminya replied:
I would use `UInt8` instead 🐠

KristofferC commented:
> This is ready to get merged.

Seems like all CI failed though.

KristofferC commented on base/dict.jl:
````diff
@@ -68,12 +68,65 @@ Dict{String,Int64} with 2 entries:
   "A" => 1
 ```
+
+After creation of a Dict you can use `d[x]` for getting the values and `d[x]=y` for changing the values.
````
This could perhaps use https://github.com/JuliaLang/julia/pull/34226 because of the docstring is getting very long now.

KristofferC commented on base/dict.jl:
````diff
@@ -68,12 +68,65 @@ Dict{String,Int64} with 2 entries:
+
+After creating an empty dictionary, you can fill it via a for loop:
+```jldoctest dictexamples
+mydict = Dict{Any, Any}()
````
`julia>` for doctests?

KristofferC replied:
Also, why specify `{Any, Any}`? Isn't it better to use a typed dict or alternatively, just write `Dict()` (which was documented above)

KristofferC commented on base/dict.jl:
````diff
@@ -68,12 +68,65 @@ Dict{String,Int64} with 2 entries:
+```
+
+To loop through key and value pairs use:
+```julia
````
Should be a doctest? Also should be a named doctest to be able to access the variables in the other doctests.

KristofferC commented on base/dict.jl:
```diff
@@ -68,12 +68,65 @@ Dict{String,Int64} with 2 entries:
+
+animals = ["cat", "fish", "elephant"]
+for (i, x) in enumerate(animals)
+   mydict[i] = [x, rand(UInt)]
```
An array seems like a bad data structure here. Also, Why use a dict if you just store the consequtive indices 1:x in them? Could just use an array then?

aminya replied:
Oh right! I should use `x` instead of `i`.

KristofferC commented on base/dict.jl:
```diff
@@ -68,12 +68,65 @@ Dict{String,Int64} with 2 entries:
+julia> Dict{Int64,Any}()  # type specified
+Dict{Int64,Any} with 0 entries
+
+julia> dict = Dict()  # type not specified
```
Why is this named `dict` but the one above not named?

KristofferC commented on base/dict.jl:
````diff
@@ -68,12 +68,65 @@ Dict{String,Int64} with 2 entries:
+
+You can use [Generator](https://docs.julialang.org/en/v1/manual/arrays/#Generator-Expressions-1) and [Comprehension][https://docs.julialang.org/en/v1/manual/arrays/#Comprehensions-1] syntax to create dictionaries:
+
+```jldoctest dictexamples
````
No need to name this doctest?

KristofferC commented on base/dict.jl:
````diff
@@ -68,12 +68,65 @@ Dict{String,Int64} with 2 entries:
+
+To loop through key and value pairs use:
+```julia
+>julia for (key, value) in mydict
````
`>julia`?

KristofferC commented on base/dict.jl:
````diff
@@ -493,6 +546,12 @@ julia> get(d, "a", 3)
 3
 ```
+
+Using `get()` is the same as using `d[x]` syntax when the key exists in the collection:
````
I don't really see what extra information this adds that isn't in the docstring above.

KristofferC commented on base/set.jl:
````diff
@@ -17,6 +17,37 @@ Set() = Set{Any}()
+Set(["cat","fish","dog","bird"])
+```
+
+You can let Julia guess the type:
````
"Guess" seems like the wrong word to use here.

KristofferC commented on base/set.jl:
````diff
@@ -17,6 +17,37 @@ Set() = Set{Any}()
+```
+
+You can let Julia guess the type:
+```jldoctest setexamples
````
No need to name this doctest?

KristofferC commented on base/set.jl:
```diff
@@ -17,6 +17,37 @@ Set() = Set{Any}()
+julia> animals = Set(["cat","fish","dog","bird"])
+Set(["cat","fish","dog","bird"])
+
+julia>typeof(animals)
```
Space missing?

KristofferC commented on doc/src/base/collections.md:
```diff
@@ -235,6 +236,56 @@ Partially implemented by:
   * [`Iterators.Pairs`](@ref)
 
 ## Set-Like Collections
+A [`Set`](@ref) is a collection of elements, just like an array or dictionary, with no duplicated elements.
```
It is a bit different from arrays or dictionaries though, since `Set`s have no keys, or?

KristofferC commented on doc/src/base/collections.md:
````diff
@@ -235,6 +236,56 @@ Partially implemented by:
+Set(["yellow", "blue", "green", "red"])
+```
+
+You can use [`push!`](@Ref) to add elements to a set:
````
Should be `@ref`?

KristofferC commented on doc/src/base/collections.md:
````diff
@@ -235,6 +236,56 @@ Partially implemented by:
+Set(["indigo", "yellow", "orange", "blue", "violet", "green", "black", "red"])
+```
+The intersection of two sets is the set that contains every element that belongs to both sets:
+```julia
````
Should be a (named) doctest?

KristofferC commented on doc/src/base/collections.md:
````diff
@@ -235,6 +236,56 @@ Partially implemented by:
+true
+```
+
+There are some standard operations you can do with sets, namely find their `union`, `intersection`, and `difference`, with the functions, [`union`](@Ref), [`intersect`](@Ref), and [`setdiff`](@Ref):
````
This seems to mostly repeat the docstrings, so I wonder how much value there is to have a secondary set of examples for e.g. `intersect` and `union`.

aminya replied:
I treated this like an all-in-one manual for `Set` separate from docstrings.

aminya replied:
Now that I saw the docstrings are provided in the `collections.md`, I decided to move the examples inside the docstrings themselves.

KristofferC commented on doc/src/base/collections.md:
````diff
@@ -235,6 +236,56 @@ Partially implemented by:
+true
+```
+
+There are some standard operations you can do with sets, namely find their `union`, `intersection`, and `difference`, with the functions, [`union`](@Ref), [`intersect`](@Ref), and [`setdiff`](@Ref):
````
I don't think the first set of `union`, `intersection` and `difference` should be code quoted since it isn't code (there isn't a `difference` function, for example).

KristofferC commented on doc/src/base/collections.md:
````diff
@@ -235,6 +236,56 @@ Partially implemented by:
+true
+```
+or using infix notation
+```julia
````
named doctest?

KristofferC commented on doc/src/base/collections.md:
````diff
@@ -235,6 +236,56 @@ Partially implemented by:
+You can't use [`pushfirst!`](@Ref), because that works only for things that have a concept of "first", like arrays.
+
+To see if something is in the set, you can use [`in`](@Ref):
+```julia
````
named doctest?

aminya committed:
Change dynamic dict filling example

Fixes https://github.com/JuliaLang/julia/pull/33936#discussion_r372377664
Fixes https://github.com/JuliaLang/julia/pull/33936#discussion_r372378255
Fixes https://github.com/JuliaLang/julia/pull/33936#discussion_r372375397
Fixes https://github.com/JuliaLang/julia/pull/33936#discussion_r372379252
Fixes https://github.com/JuliaLang/julia/pull/33936#discussion_r372378044

aminya committed:
Fix collections.md problems

Fixes https://github.com/JuliaLang/julia/pull/33936#discussion_r372384126
Fixes https://github.com/JuliaLang/julia/pull/33936#discussion_r372381479
Fixes https://github.com/JuliaLang/julia/pull/33936#discussion_r372381117
Fixes https://github.com/JuliaLang/julia/pull/33936#discussion_r372384252
Fixes https://github.com/JuliaLang/julia/pull/33936#discussion_r372384218
Fixes https://github.com/JuliaLang/julia/pull/33936#discussion_r372382984

aminya committed:
Fix dict docstring

Fixes https://github.com/JuliaLang/julia/pull/33936#discussion_r372377220
Fixes https://github.com/JuliaLang/julia/pull/33936#discussion_r372378562

using extended help for Dict

Fix get docstring

Fixes https://github.com/JuliaLang/julia/pull/33936#discussion_r372379744

dict.jl fixes

Fixes
https://github.com/JuliaLang/julia/pull/33936#discussion_r373671388

Fixes
https://github.com/JuliaLang/julia/pull/33936#discussion_r373674619

Changing Dict dynamic filling description

aminya committed:
Fixing set docstring

Fixes https://github.com/JuliaLang/julia/pull/33936#discussion_r372374989
Fixes https://github.com/JuliaLang/julia/pull/33936#discussion_r372380186
Fixes https://github.com/JuliaLang/julia/pull/33936#discussion_r372380392
Fixes https://github.com/JuliaLang/julia/pull/33936#discussion_r372380462

Set filtering/looping example

Moving union, intersect, setdiff examples to docstrings

Fixes
https://github.com/JuliaLang/julia/pull/33936#discussion_r372383584

AbstractSet fixes

Fixes
https://github.com/JuliaLang/julia/pull/33936#discussion_r373669561
Fixes
https://github.com/JuliaLang/julia/pull/33936#discussion_r373670165
Fixes
https://github.com/JuliaLang/julia/pull/33936#discussion_r373670360
Fixes
https://github.com/JuliaLang/julia/pull/33936#discussion_r373670399
Fixes
https://github.com/JuliaLang/julia/pull/33936#discussion_r373671096
Fixes
https://github.com/JuliaLang/julia/pull/33936#discussion_r373671149
Fixes
https://github.com/JuliaLang/julia/pull/33936#discussion_r373671188

set.jl fixes

Fixes
https://github.com/JuliaLang/julia/pull/33936#discussion_r373675780

Fixes
https://github.com/JuliaLang/julia/pull/33936#discussion_r373678473

Fixes
https://github.com/JuliaLang/julia/pull/33936#discussion_r373678630

Fix setdiff

aminya commented:
> 
> 
> > This is ready to get merged.
> 
> Seems like all CI failed though.

Do you know why it fails?

aminya committed:
collections.md fixes

Fixes
https://github.com/JuliaLang/julia/pull/33936#discussion_r373677105

Fixes
https://github.com/JuliaLang/julia/pull/33936#discussion_r373679117

jonas-schulze commented on doc/src/base/collections.md:
```diff
@@ -236,7 +236,9 @@ Partially implemented by:
-A [`Set`](@ref) is a collection of elements with no duplicated elements and keys/indices. So in contrast to other types of collection, in a `Set` you can have only one of each element, and the order of elements isn't important (whereas an array can have multiple copies of an element and their order is remembered).
+A [`Set`](@ref) is a collection of elements with no duplicated elements and keys/indices.
+
+So in contrast to `Arrays`, in a `Set` you can have only one of each element, and the order of elements isn't important (whereas an array can have multiple copies of an element and their order is remembered). As a result, iterating over `Set` elements is slower than `Arrays`.
```
The last sentence sounds like you were explaining some causality, which is IMO misleading. The performance depends on the specific implementation of a set, which for example could be a “plain” hash map (then iteration would most likely be a lot slower), some type of search tree (if the elements/hashes are ordered; then iteration would be about as fast as an array) or a hybrid structure equipped with an array of the unique elements (then iteration would be exactly as fast as ... an array). I would remove that last sentence.

jonas-schulze replied:
Also, I am generally not a big fan of single sentence paragraphs. So I would remove that whole commit.

aminya replied:
Current `Set` implementation is like this, so the sentence is true:

https://julialang.slack.com/archives/CF53T1DU4/p1580368042007300
> A `Set` stores the memory like a Dict, (i.e. a hash table), scattered all over the place, with empty spaces. It must jump around memory, often multiple times to fetch an element. And it must do a few simple checks every time it jumps

jonas-schulze replied:
I am not saying that you are wrong. My point is that if the documentation is based on an interface and not about a particular implementation, it should not state any properties that only depend on the implementation; at least not without referring to (and explaining) that implementation.

aminya replied:
The documentation is about the particular implementation here (not a general interface). If the implementation changes in the future, we need to update this.

mortenpi commented on base/dict.jl:
````diff
@@ -59,21 +59,77 @@ Given a single iterable argument, constructs a [`Dict`](@ref) whose key-value pa
+To loop through key and value pairs use:
+```jldoctest dictexamples
+julia> for (key, value) in animaldict
+    println("$key are in group $(value[1]), and we have $(value[2]) number of them")
````
To fix the compilation I think you need to escape these, as they are interpreted as string interpolations at the moment:

```suggestion
    println("\$key are in group \$(value[1]), and we have \$(value[2]) number of them")
```

aminya replied:
Oh yeah!! Thanks

jonas-schulze commented on base/dict.jl:
```diff
@@ -59,21 +59,77 @@ Given a single iterable argument, constructs a [`Dict`](@ref) whose key-value pa
+julia> animals = ["cats", "fishes", "elephants"];
+
+julia> for (i, x) in enumerate(animals)
+   animaldict[x] = [i, rand(UInt8)]
```
You have to chose something deterministic as a value. Otherwise your doctest will output something different each time it is run, making CI fail.

aminya replied:
Thanks! Fixed in https://github.com/JuliaLang/julia/pull/33936/commits/0b3049d415fc676d33de672cc2506ca630f0c295

KristofferC commented on base/abstractset.jl:
````diff
@@ -40,6 +42,15 @@ Set{Int64} with 3 elements:
+julia> union(A, B)
+Set(["yellow", "orange", "blue", "black"])
+```
+`union` of two `Set` is another `Set`
````
This can be left out since it is shown from the example just above.

KristofferC commented on base/abstractset.jl:
````diff
@@ -40,6 +42,15 @@ Set{Int64} with 3 elements:
   3
   1
 ```
+
````
There is usually a `#Examples` header before the examples.

KristofferC commented on base/abstractset.jl:
```diff
@@ -98,8 +109,9 @@ end
-Maintain order with arrays.
+Construct the intersection of sets. Maintain order with arrays.
+
+The union of two set `A` and `B` is a set that contains every element that belongs to both `A` and `B`.
```
This should not be about union?

KristofferC commented on base/abstractset.jl:
```diff
@@ -98,8 +109,9 @@ end
 
-Construct the intersection of sets.
-Maintain order with arrays.
+Construct the intersection of sets. Maintain order with arrays.
```
In the doctoring for `union` you changed the first sentence here. Would be good to be consistent.

aminya replied:
I am not sure what you mean here.

KristofferC commented on base/abstractset.jl:
````diff
@@ -116,6 +128,14 @@ julia> intersect(Set([1, 2]), BitSet([2, 3]))
+julia> intersect(A, B)
+Set(["orange"])
+```
+`intersect` of two `Set` is another `Set`
````
Same with this as above.

KristofferC commented on base/abstractset.jl:
````diff
@@ -116,6 +128,14 @@ julia> intersect(Set([1, 2]), BitSet([2, 3]))
+
+```jldoctest
+julia> A = Set(["yellow", "orange"]); B = Set(["blue", "black", "orange"]);
+julia> intersect(A, B)
````
Missing empty line

KristofferC commented on base/abstractset.jl:
````diff
@@ -116,6 +128,14 @@ julia> intersect(Set([1, 2]), BitSet([2, 3]))
 Set{Int64} with 1 element:
   2
 ```
+
````
`# Examples`

KristofferC commented on base/abstractset.jl:
```diff
@@ -145,13 +165,26 @@ intersect!(s::AbstractSet, itr) =
 Construct the set of elements in `s` but not in any of the iterables in `itrs`.
 Maintain order with arrays.
 
+Difference of set `A` and `B` is a set that contains the elements that are in `A`, but not in `B`.
```
Isn't this pretty much exactly what the first sentence in the docstring say? Also, the word "Difference" seems to not occur anywhere else here which makes it unclear what it refers to.

aminya replied:
Is the new one better?
https://github.com/JuliaLang/julia/pull/33936/files#diff-69a0054b980f1b009ae561651ad29c54R165

Because the difference is usually a thing between two things, I see this explanation useful.

KristofferC commented on base/abstractset.jl:
````diff
@@ -145,13 +165,26 @@ intersect!(s::AbstractSet, itr) =
+
+```jldoctest
+julia> A = Set(["yellow", "orange"]); B = Set(["blue", "black", "orange"]);
+julia> setdiff(A, B)
````
Missing newline.

KristofferC commented on base/abstractset.jl:
````diff
@@ -145,13 +165,26 @@ intersect!(s::AbstractSet, itr) =
+julia> setdiff(B, A)
+Set(["blue", "black"])
+```
+`setdiff` of two `Set` is another `Set`
````
Can be removed

KristofferC commented on base/dict.jl:
````diff
@@ -59,21 +59,88 @@ Given a single iterable argument, constructs a [`Dict`](@ref) whose key-value pa
 # Examples
-```jldoctest
-julia> Dict([("A", 1), ("B", 2)])
+```jldoctest dictexamples
````
Doesn't need to be a named dockets since the next one has the same name and overwrites `d`?

KristofferC commented on base/dict.jl:
```diff
@@ -480,8 +547,7 @@ end
 
-Return the value stored for the given key, or the given default value if no mapping for the
-key is present.
+Return the value stored for the given key, or the given default value if no mapping for the key is present.
```
Was it intentional to put this on the same line? It is a bit long now.

KristofferC commented on base/dict.jl:
````diff
@@ -493,6 +559,12 @@ julia> get(d, "a", 3)
 3
 ```
+
+Using `get()` is the same as using `d[x]` syntax when the key exists in the collection. However, when the key doesn't exist, `get` returns the default value while `d[x]` throws an error.
````
I don't really understand this part. Does it say anything that the first sentence doesn't?

aminya replied:
`d[x]` returns an error but `get` gives the default value.

KristofferC commented on base/set.jl:
````diff
@@ -17,6 +17,37 @@ Set() = Set{Any}()
+Set(["fish", "cat", "bird", "dog"])
+```
+
+You can let Julia infer the type:
````
I think sentences like this are better written like

> If the element type is omitted it is computed by intersecting the types of the arguments.

or something like that.

KristofferC commented on doc/src/base/collections.md:
```diff
@@ -235,6 +236,49 @@ Partially implemented by:
 ## Set-Like Collections
+A [`Set`](@ref) is a collection of elements with no duplicated elements and keys/indices.
+
+So in contrast to `Arrays`, in a `Set` you can have only one of each element, and the order of elements isn't important (whereas an array can have multiple copies of an element and their order is remembered). As a result, iterating over `Set` elements is slower than `Arrays`.
```
I think the part about iterating being slower can be removed. There are `AbstractArray`s have the same properties as `Array` but are slower to iterate then `Set`.

KristofferC replied:
This line is also quite long.

jonas-schulze replied:
See also https://github.com/JuliaLang/julia/pull/33936#discussion_r373215632

KristofferC reviewed (commented):
This is getting better, might need some more iterations but definitely getting there.

A lot of the added sentences start with "you" or talk directly to the author "You can", "You need" etc. This is a bit different from the voice in most of the other docstrings which is more directly talking about the thing it documents (the docstrings aren't really tutorials). I think trying to keep the voice similar to the other doctstrings would improve them quite a bit. 

It would be good to try keep the lines from getting too long as well.

KristofferC commented on base/set.jl:
````diff
@@ -17,6 +17,37 @@ Set() = Set{Any}()
+Set{String}
+```
+
+You can create an empty set using the [`Set`](@ref) constructor function and then fill it using `push!` function:
````
No need to ref `Set` multiple times.

This sentence could avoid using "you" and be written something like

> A common way of incrementally creating a `Set` is by first creating an empty set with `Se()` and then add the elements using the `push!` function.

KristofferC commented on base/set.jl:
````diff
@@ -17,6 +17,37 @@ Set() = Set{Any}()
+julia> push!(int_container, 2)
+Set([2])
+```
+```jldoctest
````
Could be merged with the doctest above.

KristofferC commented on doc/src/base/collections.md:
````diff
@@ -235,6 +236,49 @@ Partially implemented by:
+green
+red
+```
+You can also [`filter`](@ref) some of the elements:
````
Quite a lot of "You can", "You can", You can't" in this section.

KristofferC replied:
In general, I am not sure if giving examples of many of the functions on `Set` are so useful here since it duplicated the ones in the docstrings. If anything, something like

> common functions used on a `Set` are `in`[@ref], `filter`[@ref] ...

could be done, and then the examples are in those docstrings.

aminya replied:
> 
> 
> In general, I am not sure if giving examples of many of the functions on `Set` are so useful here since it duplicated the ones in the docstrings. If anything, something like
> 
> > common functions used on a `Set` are `in`[@ref], `filter`[@ref] ...
> 
> could be done, and then the examples are in those docstrings.

These functions are not given in the following docstrings, and so I thought they might be useful to have them here.

aminya commented:
> 
> 
> This is getting better, might need some more iterations but definitely getting there.
> 

It got harder than I thought. Probably that's why people don't contribute to documentation much 😄 .

> A lot of the added sentences start with "you" or talk directly to the author "You can", "You need" etc. This is a bit different from the voice in most of the other docstrings which is more directly talking about the thing it documents (the docstrings aren't really tutorials). I think trying to keep the voice similar to the other doctstrings would improve them quite a bit.

I like using active voices, and I see it as a more friendly style of writing. But since the documentation should be written in a cohesive style, I changed the sentences.

KristofferC commented:
> It got harder than I thought. Probably that's why people don't contribute to documentation much 😄 .

I wouldn't say that is true. There are PRs all the time for the docs :). But a PR that adds 200 lines will likely need to be iterated on quite a bit more than one that just does a small change.

aminya commented:
Doctest is giving an incorrect result. The error message doesn't make sense.
It says:
```julia
�� Expected output:
��� 
���   println(color)
��� end
��� yellow
��� blue
��� green
��� red
��� 
���   diff =
���    Warning: Diff output requires color.
���      println(color)
���    end
���    yellow
���    blue
���    green
���    redERROR: syntax: incomplete: premature end of input
```

mortenpi commented on doc/src/base/collections.md:
````diff
@@ -235,6 +236,60 @@ Partially implemented by:
+```jldoctest set_collections
+julia> for color in colors
+  println(color)
+end
````
This doctest is badly indented. It should be:

```
julia> for color in colors
         println(color)
       end

aminya replied:
Thank you! I was confused by this.

aminya committed:
Fixing output of code in docstrings

Using Julia 1.5 to get the examples output

Indentation fix

https://github.com/JuliaLang/julia/pull/33936#discussion_r373895887

Fixing jldoctests

KristofferC commented on base/dict.jl:
````diff
@@ -493,6 +561,12 @@ julia> get(d, "a", 3)
 3
 ```
+
+Using `get()` is the same as using `d[x]` syntax when the key exists in the collection. However, when the key doesn't exist, `get` returns the default value while `d[x]` throws an error.
````
This still just seems like a rewrite of the section above the example above this. I think it can be removed.

aminya replied:
The section above doesn't say the difference between `d[x]` and `get()`

KristofferC commented on base/set.jl:
````diff
@@ -17,6 +17,46 @@ Set() = Set{Any}()
+
+# Examples
+```jldoctest
+julia> animals = Set{String}(["cat","fish","dog","bird"])
````
```suggestion
julia> animals = Set{String}(["cat", "fish", "dog", "bird"])
```

KristofferC commented on base/set.jl:
````diff
@@ -17,6 +17,46 @@ Set() = Set{Any}()
+  "dog"
+```
+
+If the element type is not given, it is computed by intersecting the types of the arguments:
````
Would be good to have an example where not all elements are of the same type here, like

```
julia> Set([2, 1.0, Int8(3)])
Set{Float64} with 3 elements:
  2.0
  3.0
  1.0
```

KristofferC commented on base/set.jl:
```diff
@@ -17,6 +17,46 @@ Set() = Set{Any}()
+julia> any_container = Set() # type not specified
+Set{Any} with 0 elements
+
+julia> push!(any_container, "hi!")
```
You could perhaps `push!` a number into this one as well to show that you can have elements of different types in a `Set{Any}`.

KristofferC commented on doc/src/base/collections.md:
```diff
@@ -235,6 +236,61 @@ Partially implemented by:
   * [`Iterators.Pairs`](@ref)
 
 ## Set-Like Collections
+A [`Set`](@ref) is a collection of elements with no duplicated elements and keys/indices (unlike `Array` or `Dict`). So in a `Set` you can have only one of each element, and the order of elements isn't important.
```
I don't really get the "keys/indices" part here. A `Set` has no keys. So perhaps something like:

```suggestion
A [`Set`](@ref) is an unordered collection of distinct elements (unlike e.g. `Array` which is ordered and can have multiple equal elements).
```

aminya replied:
That part of sentence was saying `Set` doesn't have `key` or `indices`

KristofferC commented on doc/src/base/collections.md:
```diff
@@ -189,8 +189,9 @@ Dictionaries may also be created with generators. For example, `Dict(i => f(i) f
-`D[x,y]`  is equivalent to `D[(x,y)]`, i.e. it refers to the value keyed by the tuple `(x,y)`.
+for the key `x`).
+
+Multiple arguments to `D[...]` are converted to tuples; for example, the syntax `D[x,y]`  is equivalent to `D[(x,y)]`, i.e. it refers to the value keyed by the tuple `(x,y)`.
```
This still talks about the indexing of `D` so it makes sense to me to keep it in the same paragraph.

KristofferC commented on base/set.jl:
````diff
@@ -17,6 +17,46 @@ Set() = Set{Any}()
+
+If the element type is not given, it is computed by intersecting the types of the arguments:
+```jldoctest
+julia> animals = Set(["cat","fish","dog","bird"])
````
```suggestion
julia> animals = Set(["cat", "fish", "dog", "bird"])
```

KristofferC commented on doc/src/base/collections.md:
````diff
@@ -235,6 +236,61 @@ Partially implemented by:
+  "red"
+```
+
+Since sets don't store repeated elements, `push!` will not have any effect for pushing the elements that are already in the set. Also because `Set` doesn't have a concept of "first" (unlike `Array`), using [`pushfirst!`](@ref) to insert an item at the beginning of a set will result in an error.
````
I feel like this whole paragraph can be removed. This is mostly rehashing things that have already been mentioned.

aminya replied:
Here I am describing the context of `push!`.

KristofferC commented on base/dict.jl:
```diff
@@ -68,12 +68,80 @@ Dict{String,Int64} with 2 entries:
+
+# Extended help
+
+You can use [Generator](https://docs.julialang.org/en/v1/manual/arrays/#Generator-Expressions-1) and [Comprehension][https://docs.julialang.org/en/v1/manual/arrays/#Comprehensions-1] syntax to create dictionaries:
```
This seems like a restatment of the first part of this docstring:

> Given a single iterable argument, constructs a [`Dict`](@ref) whose key-value pairs
>  are taken from 2-tuples `(key,value)` generated by the argument.

aminya replied:
No, it is not. `Dict(x => x^2 for x = 0:1:4)` is different from `Dict([x => x^2 for x = 0:1:4])`

aminya replied:
Also, now that I am thinking, the location of the 1st and 2nd examples should be changed. The 2nd example is `2X` faster since it directly makes the `Dict`.
```julia

# Examples
```jldoctest dictexamples
julia> Dict([("A", 1), ("B", 2)])
Dict{String,Int64} with 2 entries:
  "B" => 2
  "A" => 1
```

Alternatively, a sequence of pair arguments may be passed.

```jldoctest dictexamples
julia> d = Dict("A"=>1, "B"=>2)
Dict{String,Int64} with 2 entries:
  "B" => 2
  "A" => 1
```

```

KristofferC commented on base/dict.jl:
````diff
@@ -68,12 +68,80 @@ Dict{String,Int64} with 2 entries:
+Dict{Any,Any} with 0 entries
+```
+
+After creating an empty dictionary, fill it via a for loop:
````
This is the doctstring for the constructor `Dict()`. Now we are talking about modyfing an existing dict which feels out of scope for this docstring.

aminya replied:
This is in the Extended help section of `Dict`. `Dict` doesn't have a manual like [`Array` does](https://docs.julialang.org/en/v1/manual/arrays/), so this docstring is the best place for talking about modifying a Dict.

aminya committed:
Formatting + Rewording

https://github.com/JuliaLang/julia/pull/33936#discussion_r381895269
https://github.com/JuliaLang/julia/pull/33936#discussion_r381895387
https://github.com/JuliaLang/julia/pull/33936#discussion_r381895748
https://github.com/JuliaLang/julia/pull/33936#discussion_r381900223
https://github.com/JuliaLang/julia/pull/33936/files#r381901387

aminya commented:
General note about this PR: 
This is a documentation for those who are not familiar with Julia, and may not infer all the features and capabilities from a single sentence. So having more description and example doesn't harm the user, but instead gives more knowledge.

As an example see this documentation in Python for Dictionaries:
https://docs.python.org/3/tutorial/datastructures.html#dictionaries

c42f commented:
Looking at the history of this PR, the discussion here has been very long for what I'd naively expect should be a simple change. Personally I glanced over these changes a few weeks ago and I had a feeling they "somehow didn't fit in", but couldn't express why.

However I was just thinking about the discussion at the end of https://github.com/JuliaLang/julia/pull/34226 and when I thought again of this issue it became clear what the problem is: @aminya is trying to write *Tutorial* documentation — in the sense of https://www.divio.com/blog/documentation (this is a must-read for anyone interested in good documentation) — but the Julia documentation currently has no place for tutorials.

Currently we have:
* **Reference** documentation is contained in docstrings which are terse but complete. @aminya this is the reason you are seeing resistance for adding large amounts of examples to the docstrings. Docstrings are just not a good place for extended tutorial-like content.
* **Explanation** is contained in the main Julia user manual (there is some small amount of tutorial content mixed in but that generally doesn't fit well with the style). As explanation, the manual currently doesn't serve beginners well, because tutorial content doesn't fit in nicely.
* **Tutorials** and **How-Tos** do not have a natural home.

To resolve this in a good way I think we really need an explicit home for tutorial content in the manual which is separate from the docstrings and separate from the Explanation part of the manual. Combined with good cross linking between the sections I think we'll get the best outcome.

aminya commented:
I tried to help to improve the documentation of Julia. But there is resistance for adding manual, extra information, examples, and things that are specifically useful for beginners. I had to argue with people even for documentation or for one sentence.

Similar behavior here: 
https://github.com/FluxML/Flux.jl/pull/985

StefanKarpinski commented:
> The last time that the doc project was touched was in 2017. This shows the status pretty much.
> https://github.com/JuliaLang/julia/projects/8

The only thing that shows is that we don't use the projects feature.

Doc improvements are very much appreciated and merged all the time without issue. In this case, a lot of feedback was given and you effectively rejected large amounts of it. If you won't accept feedback on a PR, then it may not get merged and the person who gave that feedback may get tired and go away.

I think the basic issue with this PR is that it adds a _lot_ of content to doc strings, which are supposed to be fairly minimal reference-style explanations of how specific APIs work. A lot of the feedback which was rejected was pushing the changes in that direction. With a more robust support for short versus extended docstrings (see https://github.com/JuliaLang/julia/pull/34226), having more and longer examples might be fine, but as it is there's a tradeoff between being more "friendly" versus overwhelming people with too many examples and details.

If you were to separate this into the manual additions, which are less controversial, and some minimal doc string improvements, that PR could be merged quickly and the addition of lots of examples to doc strings could wait on the resolution of how to handle extended docs.

ViralBShah commented:
May I suggest closing this one and opening separate smaller PRs? I have been looking at many old doc PRs and merging them. In many cases I am doing the suggested work by the reviewers myself, which takes me more time than it would have taken the author.

It's hard to figure out the right answer. The quality of reviews for even small PRs is incredibly good, that leads to higher quality documentation that will literally help everyone who reads it. However, I can also see how it can be difficult for a new contributor.
